src/SharpCompress/Common/Zip/ZipFilePart.cs

```diff
@@ -124,40 +124,46 @@ protected Stream CreateDecompressionStream(Stream stream)
             }
         }
 
-        protected Stream GetCryptoStream(Stream plainStream)
-        {
-            if ((Header.CompressedSize == 0)
-#if !NO_CRYPTO
- && ((Header.PkwareTraditionalEncryptionData != null)
-                    || (Header.WinzipAesEncryptionData != null)))
-#else
-                && (Header.PkwareTraditionalEncryptionData != null))
-#endif
-            {
-                throw new NotSupportedException("Cannot encrypt file with unknown size at start.");
-            }
-            if ((Header.CompressedSize == 0)
-                && FlagUtility.HasFlag(Header.Flags, HeaderFlags.UsePostDataDescriptor))
-            {
-                plainStream = new NonDisposingStream(plainStream); //make sure AES doesn't close    
-            }
-            else
-            {
-                plainStream = new ReadOnlySubStream(plainStream, Header.CompressedSize); //make sure AES doesn't close
-            }
-            if (Header.PkwareTraditionalEncryptionData != null)
-            {
-                return new PkwareTraditionalCryptoStream(plainStream, Header.PkwareTraditionalEncryptionData,
-                                                         CryptoMode.Decrypt);
-            }
+		protected Stream GetCryptoStream(Stream plainStream)
+		{
+			if ((Header.CompressedSize == 0) && !string.IsNullOrEmpty(Header.Password))
+			{
+				throw new NotSupportedException("Cannot encrypt file with unknown size at start.");
+			}
+
+			if ((Header.CompressedSize == 0)
+				&& FlagUtility.HasFlag(Header.Flags, HeaderFlags.UsePostDataDescriptor))
+			{
+				plainStream = new NonDisposingStream(plainStream); //make sure AES doesn't close    
+			}
+			else
+			{
+				plainStream = new ReadOnlySubStream(plainStream, Header.CompressedSize); //make sure AES doesn't close
+			}
+
+			if (FlagUtility.HasFlag(Header.Flags, HeaderFlags.Encrypted))
+			{
+				switch (Header.CompressionMethod)
+				{
+					case ZipCompressionMethod.None:
+					case ZipCompressionMethod.Deflate:
+					case ZipCompressionMethod.Deflate64:
+					case ZipCompressionMethod.BZip2:
+					case ZipCompressionMethod.LZMA:
+					case ZipCompressionMethod.PPMd:
+						return new PkwareTraditionalCryptoStream(plainStream, Header.ComposeEncryptionData(plainStream), CryptoMode.Decrypt);
+
+					case ZipCompressionMethod.WinzipAes:
 #if !NO_FILE
-            if (Header.WinzipAesEncryptionData != null)
-            {
-                //only read 10 less because the last ten are auth bytes
-                return new WinzipAesCryptoStream(plainStream, Header.WinzipAesEncryptionData, Header.CompressedSize - 10);
-            }
+						return Header.WinzipAesEncryptionData != null ? new WinzipAesCryptoStream(plainStream, Header.WinzipAesEncryptionData, Header.CompressedSize - 10) : plainStream;
```

Comment:
Why do this?

I hate the tertiary operator.  It's not readable on a single line here.